File: backend/app/schemas/approval.py
"""
Approval workflow schemas for API validation and serialization.

Module 7: Multi-Stage Approval Workflow
"""
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field, ConfigDict


# Enums
class ApprovalStageEnum(str):
    """Approval stage enumeration."""
    inspector = "inspector"
    engineer = "engineer"
    rbi = "rbi"
    team_leader = "team_leader"


class WorkflowStatusEnum(str):
    """Workflow status enumeration."""
    pending = "pending"
    in_progress = "in_progress"
    approved = "approved"
    rejected = "rejected"
    cancelled = "cancelled"


class StageStatusEnum(str):
    """Stage status enumeration."""
    pending = "pending"
    in_review = "in_review"
    approved = "approved"
    rejected = "rejected"
    skipped = "skipped"


class CommentTypeEnum(str):
    """Comment type enumeration."""
    approval = "approval"
    rejection = "rejection"
    request_changes = "request_changes"
    general = "general"


# Approval Action Schemas
class ApprovalActionRequest(BaseModel):
    """Schema for approval or rejection action."""
    comments: Optional[str] = Field(None, description="Comments for the action")

    model_config = ConfigDict(from_attributes=True)


class RejectionRequest(BaseModel):
    """Schema for rejection request."""
    comments: str = Field(..., min_length=1, description="Rejection reason (required)")

    model_config = ConfigDict(from_attributes=True)


# Response Schemas
class ApprovalCommentResponse(BaseModel):
    """Schema for approval comment response."""
    id: int
    stage_id: int
    reviewer_id: Optional[int]
    comment: str
    comment_type: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ApprovalStageResponse(BaseModel):
    """Schema for approval stage response."""
    id: int
    workflow_id: int
    stage_name: str
    stage_order: int
    reviewer_id: Optional[int]
    status: str
    comments: Optional[str]
    reviewed_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ApprovalHistoryResponse(BaseModel):
    """Schema for approval history response."""
    id: int
    workflow_id: int
    action: str
    performed_by_id: Optional[int]
    stage_name: Optional[str]
    comments: Optional[str]
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ApprovalWorkflowResponse(BaseModel):
    """Schema for approval workflow response."""
    id: int
    report_id: int
    current_stage: Optional[str]
    status: str
    created_at: datetime
    updated_at: datetime
    stages: List[ApprovalStageResponse] = []
    history: List[ApprovalHistoryResponse] = []

    model_config = ConfigDict(from_attributes=True)


# List Response with Extended Information
class PendingApprovalItem(BaseModel):
    """Schema for pending approval list item with extended information."""
    workflow_id: int
    report_id: int
    report_number: str
    inspection_id: int
    asset_id: Optional[int] = None
    asset_name: Optional[str] = None
    inspection_date: Optional[datetime] = None
    inspection_type: Optional[str] = None
    inspector_id: Optional[int] = None
    inspector_name: Optional[str] = None
    findings_count: int = 0
    critical_findings: int = 0
    high_findings: int = 0
    medium_findings: int = 0
    low_findings: int = 0
    current_stage: Optional[str] = None
    workflow_status: str
    submitted_at: Optional[datetime] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PendingApprovalsResponse(BaseModel):
    """Schema for pending approvals list response."""
    items: List[PendingApprovalItem]
    total: int
    page: int = 1
    page_size: int = 20

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/approval.ts
import { BaseEntity } from './index';

export type ApprovalStage = 'inspector' | 'engineer' | 'rbi' | 'team_leader';
export type ApprovalStatus = 'pending' | 'in_progress' | 'approved' | 'rejected' | 'cancelled';
export type StageStatus = 'pending' | 'in_review' | 'approved' | 'rejected' | 'skipped';
export type CommentType = 'approval' | 'rejection' | 'request_changes' | 'general';
export type Severity = 'critical' | 'high' | 'medium' | 'low';

export interface ApprovalComment extends BaseEntity {
  stage_id: number;
  reviewer_id: number;
  comment: string;
  comment_type: CommentType;
}

export interface ApprovalStageRecord extends BaseEntity {
  workflow_id: number;
  stage_name: string;
  stage_order: number;
  reviewer_id?: number;
  status: StageStatus;
  comments?: string;
  reviewed_at?: string;
}

export interface ApprovalHistory extends BaseEntity {
  workflow_id: number;
  action: string;
  performed_by_id?: number;
  stage_name?: string;
  comments?: string;
}

export interface ApprovalWorkflow extends BaseEntity {
  report_id: number;
  current_stage?: ApprovalStage;
  status: ApprovalStatus;
  stages: ApprovalStageRecord[];
  history: ApprovalHistory[];
}

export interface ApprovalAction {
  action: 'approve' | 'reject';
  comments?: string;
}

export interface PendingApprovalItem {
  workflow_id: number;
  report_id: number;
  report_number: string;
  inspection_id: number;
  asset_id?: number;
  asset_name?: string;
  inspection_date?: string;
  inspection_type?: string;
  inspector_id?: number;
  inspector_name?: string;
  findings_count: number;
  critical_findings: number;
  high_findings: number;
  medium_findings: number;
  low_findings: number;
  current_stage?: ApprovalStage;
  workflow_status: ApprovalStatus;
  submitted_at?: string;
  created_at: string;
}
